File: examples/governance_static_go/domain/close_incident.go
package domain

import (
	"time"
	"fmt"
)

type CloseIncident struct {
	IncidentId string `json:"incident_id"`
}

func (c CloseIncident) CommandName() string { return "CloseIncident" }

func (c CloseIncident) Execute(repo IncidentRepository) (*Incident, *ClosedIncident, error) {
	existing, err := repo.Find(c.IncidentId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("Incident not found: %s", c.IncidentId)
	}
	if existing.Status != "resolved" {
		return nil, nil, fmt.Errorf("cannot CloseIncident: Incident is in %s state", existing.Status)
	}
	existing.Status = "closed"
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := ClosedIncident{
		AggregateID: existing.ID,
		IncidentId: c.IncidentId,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}

File: examples/governance/operations_domain/lib/operations_domain/incident/commands/close_incident.rb
module OperationsDomain
  class Incident
    module Commands
      class CloseIncident
        include Hecks::Command
        emits "ClosedIncident"

        attr_reader :incident_id

        def initialize(incident_id: nil)
          @incident_id = incident_id
        end

        def call
          existing = repository.find(incident_id)
          if existing
            unless existing.status == "resolved"
              raise Hecks::Error, "Cannot CloseIncident: status must be 'resolved', got '#{existing.status}'"
            end
            Incident.new(
              id: existing.id,
              model_id: existing.model_id,
              severity: existing.severity,
              category: existing.category,
              description: existing.description,
              reported_by_id: existing.reported_by_id,
              reported_at: existing.reported_at,
              resolved_at: existing.resolved_at,
              resolution: existing.resolution,
              root_cause: existing.root_cause,
              status: "closed"
            )
          else
            raise Hecks::Error, "Incident not found: #{incident_id}"
          end
        end
      end
    end
  end
end
